### PR TITLE
Puts a ckeck on kernel_size=0 and raises an error if kernel_size=0 in conv2D

### PR DESCRIPTION
**Pull request by ghosalsattam**
fixes issue #37334

### CODE REVIEW

**Comment by fsx950223** [comment]
I suggest adding check in op kernel instead of python client.

**Comment by ghosalsattam** [comment]
Ok I am working on that

**Comment by ghosalsattam** [comment]
@fsx950223 Can I know why are you suggesting to add the check in the op kernel?

**Comment by fsx950223** [comment]
@ghosalsattam Other clients will face same problem.

**Comment by ghosalsattam** [comment]
@fsx950223 Am I supposed to change the python/ops/nn_ops.py file?

**Comment by fsx950223** [comment]
> @fsx950223 Am I supposed to change the python/ops/nn_ops.py file?

Sorry, I'm not a Tensorflow mentor. You should ask details from Googlers.

**Comment by ghosalsattam** [comment]
@tanzhenyu Please suggest the changes that should be made.

File: tensorflow/python/keras/layers/convolutional.py [hunk 1]
```diff
@@ -597,6 +597,17 @@ def __init__(self,
...
+        
+    self.checkProperKernel()
+        
+  def checkProperKernel(self):
+        if(isinstance(self.kernel_size,tuple)):
+    	    for i in self.kernel_size:
+    		    if(i==0):
+    			    raise ValueError("Kernel dimension cannot be zero")
+        else:
+    	    if(self.kernel_size==0):
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/convolutional.py]: share this with other conv layers?

**Comment by tanzhenyu** [comment]
Also can you make a unit test, and remove the yaml?

**Comment by ghosalsattam** [comment]
@tanzhenyu  I have tried to make the changes you suggested. Please check.

**Comment by ghosalsattam** [comment]
@tanzhenyu please suggest any more changes/bugs regarding the PR.

File: tensorflow/python/keras/layers/convolutional.py [hunk 2]
```diff
@@ -561,7 +577,7 @@ class Conv2D(Conv):
     ValueError: if `padding` is "causal".
     ValueError: when both `strides` > 1 and `dilation_rate` > 1.
   """
-
+  print("In Conv2d")
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/convolutional.py]: remove

File: tensorflow/python/keras/layers/unitTest.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+from tensorflow.keras.layers import Flatten
+from tensorflow.keras.layers import Dense
+import tensorflow as tf
+class Test(tf.test.TestCase):
+    def setUp(self):
+        super(Test, self).setUp()
+        self.classifier=tf.keras.Sequential()
+         
+
+    def testOutput(self):
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/unitTest.py]: try with assertRaisesException

File: .github/workflows/pythonpackage.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
+name: Python package
```
Comment (tanzhenyu) [inline .github/workflows/pythonpackage.yml]: why do we need this file here?

**Comment by ghosalsattam** [comment]
@tanzhenyu I have done all the changes you suggested.

**Comment by ghosalsattam** [comment]
@tanzhenyu please give a review on the changes made.

File: tensorflow/python/keras/layers/convolutional.py [hunk 5]
```diff
@@ -426,24 +426,43 @@ def __init__(self,
...
+        
+  def checkProperKernel(self,kernel_size):
+        if(isinstance(kernel_size,tuple)):
+    	    for i in kernel_size:
+    		    if(i==0):
+    			    raise ValueError("Kernel dimension cannot be zero")
+        else:
+    	    if(kernel_size==0):
+    		    raise ValueError("Kernel size cannot be zero")
+        return 0
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/convolutional.py]: this is a very weird return value....the previous implementation looks mostly good, simply just check without returning anything

Comment (ghosalsattam) [reply tensorflow/python/keras/layers/convolutional.py]: Can I use some other return value like("True")  as some kind of check needs to be performed before super(Conv1D) is called? Else the super() function will be called at any cost and the problem addressed in #37334 will still exist.

**Comment by ghosalsattam** [comment]
Done!!!

File: tensorflow/python/keras/layers/convolutional.py [hunk 6]
```diff
@@ -579,24 +600,41 @@ def __init__(self,
...
+          bias_regularizer=regularizers.get(bias_regularizer),
+          activity_regularizer=regularizers.get(activity_regularizer),
+          kernel_constraint=constraints.get(kernel_constraint),
+          bias_constraint=constraints.get(bias_constraint),
+          **kwargs)
+    
+    
+    
+        
+  def checkProperKernel(self,kernel_size):
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/convolutional.py]: why not use that in some base layer, or a regular function?

File: tensorflow/python/keras/layers/convolutional.py [hunk 7]
```diff
@@ -426,24 +426,45 @@ def __init__(self,
...
-        use_bias=use_bias,
-        kernel_initializer=initializers.get(kernel_initializer),
-        bias_initializer=initializers.get(bias_initializer),
-        kernel_regularizer=regularizers.get(kernel_regularizer),
-        bias_regularizer=regularizers.get(bias_regularizer),
-        activity_regularizer=regularizers.get(activity_regularizer),
-        kernel_constraint=constraints.get(kernel_constraint),
-        bias_constraint=constraints.get(bias_constraint),
-        **kwargs)
+          rank=1,
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/convolutional.py]: this indent doesn't seem correct.

File: tensorflow/python/keras/layers/convolutional.py (hunk 6 above)
Comment (ghosalsattam) [reply tensorflow/python/keras/layers/convolutional.py]: I had implemented the function in base class Conv.

**Comment by ghosalsattam** [comment]
I had implemented the function on base class Conv and ran the unit test
with different parameter values. It was running correctly(No indentation
error was detected). For all the conv layers, I checked both kernel_size=0
and kernel_size non-zero. It was working as expected(raising error when
kernel_size=0 was encountered).

**Comment by ghosalsattam** [comment]
@tanzhenyu is the problem solved?

**Comment by ghosalsattam** [comment]
@tanzhenyu Please give a review on the changes.

**Comment by ghosalsattam** [comment]
@tanzhenyu @rthadur ,Please provide a review.

**Comment by tanzhenyu** [comment]
Closing this since it has been fixed [here](https://github.com/tensorflow/tensorflow/commit/1e102f63964365d82d7f22402b7ba21e0e0e64fe).

**Pull request closed without merging**